enhancements/microshift/periodic-etcd-defragmentation-in-microshift.md

```diff
@@ -0,0 +1,168 @@
+
+### User Stories
+
+1. "As a Microshift Device Administrator, I want etcd to automatically defragment itself so that it doesn't run out of space, or use more space than it needs."
+
+### Goals
+
+- Etcd is automatically and periodically defragmented by a control loop that fits into the Microshift paradigm.
+
+### Non-Goals
+
+- Replicate other aspects/control loops of the OCP CEO.
+
+## Proposal
+
+All the following solutions will periodically calculate the "fragmented percentage" (db size on disk vs db in use). The frequency of this calculation and the threshold which the defragmentation occurs could be configurable (with defaults taken from OCP) by a Microshift Administrator so they can control when the defragmentation occurs.
+This configuration should also have the option to turn defragmentation completely off so that if an Administrator wants to opt-out of the potential for short disruptions, they have the ability to do so with the knowledge that etcd could run out of space during execution.
+For short-lived, light workload clusters, turning it off completely is likely not to cause problems.
+
+Additionally, it may be worth adding a pure periodic defragmentation so that an Administrator could force a defragmentation when load on the system is known to be low. 
+
+These two approaches could be combined so that the pure periodic defragmentation handles the normal day-to-day activity, while the threshold defragmentation is there if there are any sudden spikes in fragementation (caused by heavy usage).
+
+There is also the option of running defragmentation at Microshift startup (after etcd is alive and well). This would likely make the pure periodic redundant; however, we should still have the inflight threshold detection as that could help us avoid an out of space issue while the cluster is running.
+
+### Defragmentation from a goroutine in Microshift
+This solution would launch a goroutine, inside the Microshift binary, that would wait for the condition(s) to run defragmentation.
+This would allow for the logs/status of the defragmentation to appear in the same log stream as the other Microshift logs so and Administrator could see its status and if the defragmentation failed in the same place as they would be monitoring for other Microshift issues/status.
+
+This approach would easily afford the aforementioned periodic/threshold dual condition for defragmentation. It would wait on two timeout channels, one for pure periodic, one (shorter) for calculations; when one fires, defragmentation is started.
+
+#### Defragmentation from cron
+This approach would have the defragmentation condition(s) checked via a cron-job. When Microshift is installed, the cron-job(s) could be added to the tab and the system would handle the timing of the pure periodic and/or the threshold calculation.
```

Comment:
what command would `cron` use to trigger that? we don't install etcdctl
goroutine in `microshift{,-etcd}` could use etcd go client

Reply:
That's a very good point, I think this essentially kills the cron idea; since we don't have etcdctl (yet), we'll need a solution that has direct access to etcd.

Reply:
We're going to want etcdctl, or something like it, for sos. But let's see what we can do with the goroutine approach